controllers/default.py

```diff
@@ -212,16 +212,16 @@ def index():
                 db(
                     (db.user_sub_chapter_progress.user_id == auth.user.id)
                     & (db.user_sub_chapter_progress.chapter_id == chapter_label)
+                    & (db.user_sub_chapter_progress.course_name == course.course_name)
                 ).count()
                 == 0
             ):
                 db.executesql(
-                    """
-                    INSERT INTO user_sub_chapter_progress(user_id, chapter_id,sub_chapter_id, status, start_date)
-                    SELECT %s, chapters.chapter_label, sub_chapters.sub_chapter_label, -1, now()
-                    FROM chapters, sub_chapters where sub_chapters.chapter_id = chapters.id and chapters.course_id = '%s';
+                    f"""
```

Comment:
The change here is mainly using f-strings instead of the % operator, but the whole things scares me from a security standpoint. I know it's more pain, but I think **all** raw SQL with replacement strings should either be written using DAL or using the [placeholders argument of executsql](http://www.web2py.com/books/default/chapter/29/06/the-database-abstraction-layer#executesql).

Reply:
I agree with you in general.  But in this case there is no form to process, the id field is assigned by the system. course name and base course name also come from the database and are inserted with DAL.  This is one of those odd queries that I don't have any idea how to do in DAL, and the readability of this is just so much nicer.

Reply:
True -- DAL doesn't support this. Perhaps

``` Python
db.executesql("""
    INSERT INTO user_sub_chapter_progress(user_id, chapter_id,sub_chapter_id, status, start_date, course_name)
    SELECT %s, chapters.chapter_label, sub_chapters.sub_chapter_label, -1, now(), %s
    FROM chapters, sub_chapters where sub_chapters.chapter_id = chapters.id and chapters.course_id = %s;
""", placeholders=(auth.user.id, course.course_name, course.base_course))
```

Reply:
I know it safer but I think that syntax is mean and misleading for anyone who knows printf.  Why is everything regardless of type %s?  It makes no sense.  If you are going to do that why not go all the way and use %d for ints and %f for floats.

Reply:
I agree. This syntax is just terrible. However, it's the only way I know to sanitize SQL queries. Sigh. Looks like a place for some good docs about awful syntax. Very ugly...

Reply:
slightly more readable would be the `%(userid)s` route and then you can pass a dictionary of placeholders.

Reply:
Good idea, I like that!

Reply:
Looks like only 20 places in the code where we use db.executesql -- I'm going to make sure they are all sanitized. Even the ones I think are perfectly safe. 😬

Reply:
Awesome. I'll definitely sleep better at night. "After all, what could possibly go wrong..."